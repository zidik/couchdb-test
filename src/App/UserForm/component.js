import React from 'react';
import { Field, Formik } from 'formik';
import { Button, Form } from 'semantic-ui-react';

import { UserSchema } from '../Schemas/UserSchema';
import { isRequiredBySchema } from '../Schemas/schemaHelpers';
import { generateUniqueId } from '../helpers';
import { InputField } from '../InputField';

class UserForm extends React.Component {
  state = {
    enableReinitialize: true,
    user: null
  };
  setReinitialization = bool => this.setState({ enableReinitialize: bool });

  static getDerivedStateFromProps(props, state) {
    if (!state.enableReinitialize) return null;
    //Update user only, if reinitialization is enabled
    return { user: props.user };
  }

  render() {
    return (
      <Formik
        enableReinitialize={this.state.enableReinitialize}
        initialValues={this.state.user}
        onSubmit={async (updatedUser, { setSubmitting, resetForm }) => {
          await this.props.handleSubmit(this.state.user, updatedUser);
          setSubmitting(false);
          resetForm();
        }}
        validationSchema={UserSchema}
        render={props => (
          <StatefulUserForm
            {...props}
            setReinitialization={this.setReinitialization}
          />
        )}
      />
    );
  }
}

class StatefulUserForm extends React.Component {
  componentDidUpdate(prevProps, prevState, snapshot) {
    //has there been a change?
    if (this.props.dirty === prevProps.dirty) return;
    // disable reinitialization, if now dirty
    // enable reinitialization, it now clean
    this.props.setReinitialization(!this.props.dirty);
  }
  render() {
    return <RenderForm {...this.props} />;
  }
}

const RenderForm = ({ errors, touched, isSubmitting, handleSubmit }) => (
  <Form onSubmit={handleSubmit}>
    <Field
      id={generateUniqueId()}
      name="firstName"
      label="First Name"
      required={isRequiredBySchema(UserSchema, 'firstName')}
      component={InputField}
    />

    <Field
      id={generateUniqueId()}
      name="lastName"
      label="Last Name"
      required={isRequiredBySchema(UserSchema, 'lastName')}
      component={InputField}
    />

    <Field
      id={generateUniqueId()}
      name="email"
      label="E-mail"
      type="email"
      required={isRequiredBySchema(UserSchema, 'email')}
      component={InputField}
    />

    <Button type="submit" disabled={isSubmitting}>
      Submit
    </Button>
  </Form>
);

export default UserForm;
