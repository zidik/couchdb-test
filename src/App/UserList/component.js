import React from 'react';
import { List, Label } from 'semantic-ui-react';

export const UserList = ({
  users,
  handleSelect,
  handleNewUser,
  selectedUserId
}) => (
  <List selection divided verticalAlign="middle">
    {Object.values(users).map(user => (
      <UserListItem
        active={user._id === selectedUserId}
        key={user._id}
        user={user}
        handleSelect={handleSelect}
      />
    ))}
    <NewUserListItem handleNewUser={handleNewUser} />
  </List>
);
export const UserListItem = ({ user, handleSelect, active }) => (
  <List.Item
    active={active}
    onClick={() => handleSelect(user._id)}
    style={{ position: 'relative' }}
  >
    <List.Header>{`${user.firstName} ${user.lastName}`}</List.Header>
    <List.Content>{user.email}</List.Content>
    {user._conflicts && (
      <Label size="tiny" attached="bottom right" color="orange">
        Conflict
      </Label>
    )}
  </List.Item>
);
export const NewUserListItem = ({ handleNewUser }) => (
  <List.Item icon="add" onClick={handleNewUser} header="Add new user" />
);

export default UserList;
