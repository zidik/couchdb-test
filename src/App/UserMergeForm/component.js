import React from 'react';
import { Button, Icon, Table, Form } from 'semantic-ui-react';
import './style.css';
import { UserSchema } from '../Schemas/UserSchema';
import { isRequiredBySchema } from '../Schemas/schemaHelpers';
import { generateUniqueId } from '../helpers';
import { Field, Formik } from 'formik';
import { InputField } from '../InputField';

/***
 * Returns true if there exists a value inside the array that is different from others.
 * @param arr
 * @param selector selects value to compare
 */
const allNotEqual = (arr, selector = u => u) =>
  arr.map(selector).some((value, index, a) => value !== a[0]);

const UserMergeFormRow = ({ versions, label, name, selector }) => (
  <Table.Row error={allNotEqual(versions, selector)}>
    <Table.Cell>{name}</Table.Cell>
    <Table.Cell>
      <Field
        id={generateUniqueId()}
        name={name}
        required={isRequiredBySchema(UserSchema, name)}
        component={InputField}
      />
    </Table.Cell>
    {versions.map(user => (
      <Table.Cell key={user._rev}>{selector(user)}</Table.Cell>
    ))}
  </Table.Row>
);

const UserMergeForm = ({ user, versions, handleSubmit }) => (
  <Formik
    initialValues={user}
    onSubmit={async (updatedUser, { setSubmitting }) => {
      await handleSubmit(updatedUser, versions);
      setSubmitting(false);
    }}
    validationSchema={UserSchema}
    render={props => <RenderUserMergeForm {...props} versions={versions} />}
  />
);

const RenderUserMergeForm = ({ versions, handleSubmit, isSubmitting }) => (
  <Form onSubmit={handleSubmit}>
    <Table compact celled definition>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell />
          <Table.HeaderCell>Result</Table.HeaderCell>
          {versions.map((user, index) => (
            <Table.HeaderCell key={user._rev}>#{index + 1}</Table.HeaderCell>
          ))}
        </Table.Row>
      </Table.Header>

      <Table.Body>
        <UserMergeFormRow
          versions={versions}
          label="First Name"
          selector={user => user.firstName}
          name="firstName"
        />
        <UserMergeFormRow
          versions={versions}
          label="Last Name"
          selector={user => user.lastName}
          name="lastName"
        />
        <UserMergeFormRow
          versions={versions}
          label="E-mail"
          selector={user => user.email}
          name="email"
        />
      </Table.Body>

      <Table.Footer fullWidth>
        <Table.Row>
          <Table.HeaderCell />
          <Table.HeaderCell colSpan="4">
            <Button
              primary
              type="submit"
              size="small"
              icon
              labelPosition="left"
              disabled={isSubmitting}
            >
              <Icon name="shuffle" /> Merge
            </Button>
          </Table.HeaderCell>
        </Table.Row>
      </Table.Footer>
    </Table>
  </Form>
);

export default UserMergeForm;
