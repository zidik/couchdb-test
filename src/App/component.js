import React from 'react';
import { Header } from 'semantic-ui-react';
import 'semantic-ui-css/semantic.min.css';
import './style.css';
import UserList from './UserList';
import UserForm from './UserForm';
import UserMergeForm from './UserMergeForm';

const App = ({ remoteDbStatus, selectedUserIsInConflict }) => (
  <div className="App">
    <div className="App-header">
      <Header as="h1">CouchDB Test</Header>
    </div>
    <div className="sidebar panel">
      <Header as="h2">Users</Header>
      <UserList />
    </div>
    <div className="content panel">
      {selectedUserIsInConflict ? <UserMergeForm /> : <UserForm />}
    </div>
    <div className="App-footer">
      Sync status:
      {remoteDbStatus}
    </div>
  </div>
);

export default App;
