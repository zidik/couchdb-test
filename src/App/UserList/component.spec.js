import React from 'react';
import { shallow, mount, render } from 'enzyme';

import { NewUserListItem } from './component';

describe('NewUserListItem', () => {
  it('should render correctly', () => {
    const component = shallow(<NewUserListItem handleNewUser={() => {}} />);

    expect(component).toMatchSnapshot();
  });

  it('contains text "Add new user"', () => {
    const component = render(<NewUserListItem handleNewUser={() => {}} />);
    expect(component.text()).toContain('Add new user');
  });

  it('should call "handleNewUser" when clicked', () => {
    const clickFn = jest.fn();
    const component = mount(<NewUserListItem handleNewUser={clickFn} />);
    component.find('.item').simulate('click');
    expect(clickFn).toHaveBeenCalled();
  });
});
