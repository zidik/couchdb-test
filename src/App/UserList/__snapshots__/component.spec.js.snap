// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewUserListItem should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    handleNewUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "header": "Add new user",
      "icon": "add",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "header": "Add new user",
        "icon": "add",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
